### DISCUSSION TITLE
What's the *correct* way of managing env variables?

### DISCUSSION THREAD

**Discussion by DavidChouinard** [vercel/vercel#5055] [Help] [answered]
Here's our workflow for adding a new environment variable:

- Add the variable to the Vercel UI and do it 3x: separately for prod, preview and dev
- Run `vercel secrets list` to get the secret associated with the env variable
- Add the env variable + secret to `vercel.json`. Since `vercel.json` doesn't seem to have a concept of environment, use the production version of the secrets, i.e. use the prod keys in preview (why was it necessary to add the secret to the preview environment? would it even be possible to have different env variable for preview and prod?)
- If the variable needs to be available client-side, add it to `next.config.js`
- Instruct everyone working on the repo to run `vercel env pull` to get an updated `.env` file for developement

Seems quite convoluted and error-prone, just want to make sure — are we doing this right? With Netlify, all that was needed was just adding it to the UI and Netlify made sure to provide the right environment variables to all the environments transparently.

**Comment by styfle** [comment, marked as answer]
Hi @DavidChouinard 

> all that was needed was just adding it to the UI and Netlify made sure to provide the right environment variables to all the environments transparently.

Yep, that's all you need to do with Vercel!

If you're using the dashboard to add environment variables, then there are no other steps necessary. You can use `process.env.VARIABLENAMEHERE` to read your environment variable.

If you need the variable to be available client side, then you can prefix the name with `NEXT_PUBLIC_`.

Lastly, if you want to run `next dev` locally, you can run `vercel env pull .env.local` first to pull your Development Environment Variables from the cloud.

The Next.js documentation has more detailed information here: https://nextjs.org/docs/basic-features/environment-variables

**Comment by DavidChouinard** [comment]
Ah awesome, thank you @styfle. Not sure where along the way we started thinking we need all the extra complexity (docs on `vercel.json`, maybe?) but this all makes a ton of sense — thank you!
